python/docs/source/user_guide/sql/python_udtf.rst

```diff
@@ -0,0 +1,216 @@
+
+.. code-block:: python
+
+    class PythonUDTF:
+
+        def __init__(self) -> None:
+            """
+            Initialize the user-defined table function (UDTF).
+
+            This method serves as the default constructor and is called once when the
+            UDTF is instantiated on the executor side.
+            
+            Any class fields assigned in this method will be available for subsequent
+            calls to the `eval` and `terminate` methods.
+
+            Notes
+            -----
+            - This method does not accept any extra arguments.
+            - You cannot create or reference the Spark session within the UDTF. Any
+              attempt to do so will result in a serialization error.
+            """
+            ...
+
+        def eval(self, *args: Any) -> Iterator[Any]:
+            """
+            Evaluate the function using the given input arguments.
+
+            This method is required and must be implemented.
+
+            Argument Mapping:
+            - Each provided scalar expression maps to exactly one value in the
+              `*args` list.
+            - Each provided table argument maps to a pyspark.sql.Row object containing
+              the columns in the order they appear in the provided input relation.
+
+            This method is called on every input row, and can produce zero or more
+            output rows. Each element in the output tuple corresponds to one column
+            specified in the return type of the UDTF.
+
+            Parameters
+            ----------
+            *args : Any
+                Arbitrary positional arguments representing the input to the UDTF.
+
+            Yields
+            ------
+            tuple
+                A tuple representing a single row in the UDTF result relation.
+                Yield as many times as needed to produce multiple rows.
+
+            Notes
+            -----
+            - The result of the function must be a tuple representing a single row
+              in the UDTF result relation.
+            - UDTFs currently do not accept keyword arguments during the function call.
+
+            Examples
+            --------
+            >>> def eval(self, x: int, y: int) -> Iterator[Any]:
+            >>>     yield x + y, x - y
+            >>>     yield y + x, y - x
+            """
+            ...
+
+        def terminate(self) -> Iterator[Any]:
+            """
+            Called when the UDTF has processed all input rows.
+
+            This method is optional to implement and is useful for performing any
+            cleanup or finalization operations after the UDTF has finished processing
+            all rows. It can also be used to yield additional rows if needed.
+
+            Yields
+            ------
+            tuple
+                A tuple representing a single row in the UDTF result relation.
+                Yield this if you want to return additional rows during termination.
```

Comment:
should we mention here the tricky detail that you have to include a trailing comma when yielding a row of just one value (here and above in the `eval` description)?

Reply:
Yea let me add an example